### PR TITLE
Fix to return a task with a non-nullable IncomingResponseFrame from SendRequestAsync

### PR DESCRIPTION
**Pull request by bentoi**
This is a draft PR to discuss how to improve SendRequestAsync. Right now the PR updates the SendRequestAsync API to return a Task<IncomingRequestFrame> instead of Task<IncomingRequestFrame?>. In other words, the connection and the collocated request handler now return an incoming request frame for oneway requests.

I've added a static method to IncomingResponseFrame to get a shared incoming response frame task for oneways...

### CODE REVIEW

**Comment by pepone** [comment]
I prefer this with caches frames for void requests, than the nullable frames

**Comment by bentoi** [comment]
This PR is ready for review, it includes the following changes:
- caching of incoming/outgoing void return value frames
- fixed `IRequestHandler.SendRequestAsync` to return a `ValueTask<IncomingResponseFrame>` and accept an `IProgress<bool>` parameter.

File: csharp/src/Ice/CollocatedRequestHandler.cs [hunk 1]
```diff
@@ -54,65 +55,68 @@ internal class CollocatedRequestHandler : IRequestHandler
...
-                    observer?.Failed(ex.GetType().FullName ?? "System.Exception");
-                    throw;
-                }
-                finally
+                OutgoingResponseFrame outgoingResponseFrame = await task.WaitAsync(cancel).ConfigureAwait(false);
+
+                var incomingResponseFrame = new IncomingResponseFrame(
+                    _adapter.Communicator,
+                    outgoingRequest.Protocol,
+                    VectoredBufferExtensions.ToArray(outgoingResponseFrame!.Data));
```
Comment (pepone) [inline csharp/src/Ice/CollocatedRequestHandler.cs:99]: Seems we no longer need ! for `outgoingResponseFrame`

File: csharp/src/Ice/Communicator.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@
 using System.Security.Cryptography.X509Certificates;
 using System.Text;
 using System.Threading;
+using System.Threading.Tasks;
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:18]: This new using is not required

File: csharp/src/Ice/IncomingResponseFrame.cs [hunk 3]
```diff
@@ -32,6 +38,19 @@ public sealed class IncomingResponseFrame
...
+                                                                Protocol protocol,
+                                                                Encoding encoding) =>
+            _cachedVoidReturnValueFrames.GetOrAdd((communicator, protocol, encoding), key =>
+            {
+                var data = new List<ArraySegment<byte>>();
+                var stream = new OutputStream(key.Protocol.GetEncoding(), data, new OutputStream.Position(0, 0));
+                stream.WriteByte((byte)ReplyStatus.OK);
+                _ = stream.WriteEmptyEncapsulation(key.Encoding);
+                Debug.Assert(data.Count == 1);
+                return new IncomingResponseFrame(key.Communicator, key.Protocol, data[0]);
```
Comment (pepone) [inline csharp/src/Ice/IncomingResponseFrame.cs:51]: Wouldn't be simpler to keep the empty payloads and use them here to construct the frame?
```suggestion
                return new IncomingResponseFrame(
                    key.Communicator,
                    key.Protocol, 
                    key.Encoding == OldEncoding ? Ice1Definitions.EmptyResponsePayload : Ice2Definitions.EmptyResponsePayload);
```

File: csharp/src/Ice/OutgoingResponseFrame.cs [hunk 4]
```diff
@@ -79,10 +91,13 @@ public static OutgoingResponseFrame WithVoidReturnValue(Current current)
             byte[] buffer = new byte[256];
             buffer[0] = (byte)ReplyStatus.OK;
             response.Data.Add(buffer);
-            var ostr = new OutputStream(current.Protocol.GetEncoding(), response.Data, new OutputStream.Position(0, 1),
-                response.Encoding, format ?? current.Adapter.Communicator.DefaultFormat);
-            writer(ostr, value);
-            ostr.Save();
+            var stream = new OutputStream(current.Protocol.GetEncoding(),
```
Comment (pepone) [inline csharp/src/Ice/OutgoingResponseFrame.cs:94]: I think we use `ostr` and `istr` now for all Input/Output streams

File: csharp/src/Ice/Proxy.cs [hunk 5]
```diff
@@ -559,5 +538,23 @@ public static class Proxy
                 }
             }
         }
+
+        private class ProgressWrapper : IProgress<bool>
+        {
+            private readonly IProgress<bool>? _progress;
+
+            public bool IsSent { get; private set; }
```
Comment (pepone) [inline csharp/src/Ice/Proxy.cs:546]: should this be `internal` instead of `public`?

File: csharp/src/Ice/OutgoingResponseFrame.cs [hunk 6]
```diff
@@ -34,8 +39,15 @@ public sealed class OutgoingResponseFrame
...
-        public static OutgoingResponseFrame WithVoidReturnValue(Current current)
-            => new OutgoingResponseFrame(current.Protocol, current.Encoding, writeVoidReturnValue: true);
+        public static OutgoingResponseFrame WithVoidReturnValue(Current current) =>
+            _cachedVoidReturnValueFrames.GetOrAdd((current.Protocol, current.Encoding), key =>
+            {
+                var data = new List<ArraySegment<byte>>();
+                var stream = new OutputStream(key.Protocol.GetEncoding(), data, new OutputStream.Position(0, 0));
+                stream.WriteByte((byte)ReplyStatus.OK);
+                _ = stream.WriteEmptyEncapsulation(key.Encoding);
+                return new OutgoingResponseFrame(current.Protocol, current.Encoding, data);
```
Comment (bernardnormier) [inline csharp/src/Ice/OutgoingResponseFrame.cs:49]: Should be key.Protocol, key.Encoding?

File: csharp/src/Ice/IncomingResponseFrame.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/IncomingResponseFrame.cs:51]: This wouldn't be correct, and the old Empty payloads only correspond to 2 combinations (out of 4):
- ice2 with 2.0 header and 2.0 empty payload encoding
and
- ice1 with 1.0/1 header and 1.1 empty payload encoding

File: csharp/src/Ice/Proxy.cs (hunk 5 above)
Comment (bentoi) [reply csharp/src/Ice/Proxy.cs:546]: It should be internal

File: csharp/src/Ice/OutgoingResponseFrame.cs (hunk 6 above)
Comment (bentoi) [reply csharp/src/Ice/OutgoingResponseFrame.cs:49]: yes

**Pull request merged**